File: 2024/cpp/day7.cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

long int part1();
long int part2();

int main() {
  cout << part1() << endl;
  cout << part2() << endl;
}

long int part1() {
  string line;
  ifstream input("../data/input7.txt");

  map<long int, vector<int>> equations;

  if (input.is_open()) {
    while (getline(input, line)) {
      stringstream ss(line.substr(line.find(':') + 2));
      long int result = stol(line.substr(0, line.find(':')));

      vector<int> values;

      string val;
      while (ss >> val) {
        values.push_back(stol(val));
      }

      equations[result] = values;
    }
    input.close();
  }

  set<char> operators = {'+', '*'};
  long int true_sum = 0;

  for (auto keyval : equations) {
    long int result = keyval.first;
    vector<int> values = keyval.second;
    int count = values.size() - 1;
    bool found = false;

    vector<vector<char>> ops_combs;
    for (int i = 0; i < pow(operators.size(), count); i++) {
      vector<char> ops;
      int num = i;
      for (int j = 0; j < count; j++) {
        auto it = operators.begin();
        advance(it, num % operators.size());
        ops.push_back(*it);
        num /= operators.size();
      }
      ops_combs.push_back(ops);
    }

    for (auto op_comb : ops_combs) {
      long int curr = values[0];

      for (size_t i = 0; i < op_comb.size(); i++) {
        if (op_comb[i] == '+') {
          curr += values[i + 1];
        } else if (op_comb[i] == '*') {
          curr *= values[i + 1];
        }
      }

      if (curr == result) {
        found = true;
        break;
      }
    }

    if (found) {
      true_sum += result;
    }
  }

  return true_sum;
}

long int part2() {
  string line;
  ifstream input("../data/input7.txt");

  map<long int, vector<int>> equations;

  if (input.is_open()) {
    while (getline(input, line)) {
      stringstream ss(line.substr(line.find(':') + 2));
      long int result = stol(line.substr(0, line.find(':')));

      vector<int> values;

      string val;
      while (ss >> val) {
        values.push_back(stol(val));
      }

      equations[result] = values;
    }
    input.close();
  }

  set<char> operators = {'+', '*', '|'};
  long int true_sum = 0;

  for (auto keyval : equations) {
    long int result = keyval.first;
    vector<int> values = keyval.second;
    int count = values.size() - 1;
    bool found = false;

    vector<vector<char>> ops_combs;
    for (int i = 0; i < pow(operators.size(), count); i++) {
      vector<char> ops;
      int num = i;
      for (int j = 0; j < count; j++) {
        auto it = operators.begin();
        advance(it, num % operators.size());
        ops.push_back(*it);
        num /= operators.size();
      }
      ops_combs.push_back(ops);
    }

    for (auto op_comb : ops_combs) {
      long int curr = values[0];

      for (size_t i = 0; i < op_comb.size(); i++) {
        if (op_comb[i] == '+') {
          curr += values[i + 1];
        } else if (op_comb[i] == '*') {
          curr *= values[i + 1];
        } else if (op_comb[i] == '|') {
          curr = curr * (int)pow(10, (int)(log10(values[i + 1]) + 1)) +
                 values[i + 1];
        }
      }

      if (curr == result) {
        found = true;
        break;
      }
    }

    if (found) {
      true_sum += result;
    }
  }

  return true_sum;
}

File: 2024/python/day7.py
import math
from itertools import product


def part1() -> int:
    equations = {}
    with open("../data/input7.txt", "r") as inp:
        lines = inp.readlines()
        for line in lines:
            res, values = line.split(":")
            equations[int(res)] = tuple(map(int, values.split()))
    operators = ["+", "*"]
    true_sum = 0

    for res in equations:
        vals = equations[res]
        found = False
        for ops_combo in product(operators, repeat=len(vals) - 1):
            curr = vals[0]
            for i, op in enumerate(ops_combo):
                if op == "+":
                    curr = curr + vals[i + 1]
                elif op == "*":
                    curr = curr * vals[i + 1]
            if curr == res:
                found = True
                break
        if found:
            true_sum += res

    return true_sum


def part2() -> int:
    equations = {}
    with open("../data/input7.txt", "r") as inp:
        lines = inp.readlines()
        for line in lines:
            res, values = line.split(":")
            equations[int(res)] = tuple(map(int, values.split()))

    operators = ["+", "*", "||"]
    true_sum = 0

    for res in equations:
        vals = equations[res]
        found = False
        for ops_combo in product(operators, repeat=len(vals) - 1):
            curr = vals[0]
            for i, op in enumerate(ops_combo):
                if op == "+":
                    curr = curr + vals[i + 1]
                elif op == "*":
                    curr = curr * vals[i + 1]
                elif op == "||":
                    curr = curr * (10 ** math.ceil(math.log10(vals[i + 1]) + 1)) + vals[i + 1]
            if curr == res:
                found = True
                break
        if found:
            true_sum += res

    return true_sum


print(part1())
print(part2())
